### PR TITLE
⚡️ Encapsulate type instability of backend

### PR DESCRIPTION
**Pull request by blegat**
This PR uses the [function barrier technique](https://docs.julialang.org/en/v1/manual/performance-tips/index.html#kernel-functions-1) to reduce the impact of the type instability of `backend(model)` in `JuMP.add_constraint`.
As `JuMP.moi_add_constraint` needs to call `bridge_constraints` and `mode` with the MOI backend, we also need to apply the function barrier technique to these two functions.

This result in a small speedup in `speed.jl`. Compare the result of https://github.com/JuliaOpt/JuMP.jl/pull/1640 with the result with this PR:
```
P-Median(100 facilities, 100 customers, 5000 locations) benchmark:
BenchmarkTools.Trial: 
  memory estimate:  714.37 MiB
  allocs estimate:  12030567
  --------------
  minimum time:     1.692 s (52.39% GC)
  median time:      1.729 s (53.90% GC)
  mean time:        1.906 s (57.67% GC)
  maximum time:     2.298 s (64.40% GC)
  --------------
  samples:          3
  evals/sample:     1
Cont5(n=500) benchmark:
BenchmarkTools.Trial: 
  memory estimate:  402.77 MiB
  allocs estimate:  6785839
  --------------
  minimum time:     777.018 ms (40.46% GC)
  median time:      877.130 ms (44.78% GC)
  mean time:        871.167 ms (45.42% GC)
  maximum time:     955.600 ms (49.90% GC)
  --------------
  samples:          6
  evals/sample:     1
```

### CODE REVIEW

File: src/JuMP.jl [hunk 1]
```diff
@@ -279,19 +279,28 @@ MathOptInterface or solver-specific functionality.
...
-function mode(model::Model)
-    if !(backend(model) isa MOIU.CachingOptimizer)
-        return Direct
-    elseif backend(model).mode == MOIU.Automatic
-        return Automatic
-    else
-        return Manual
-    end
+mode(model::Model) = moi_mode(backend(model))
+# the type of backend(model) is unknown so we directly redirects to another
```
Comment (mlubin) [inline src/JuMP.jl]: Nit: Capitalization and missing period at the end. s/redirects/redirect/

File: src/constraints_future.jl [hunk 2]
```diff
@@ -0,0 +1,377 @@
+#  Copyright 2017, Iain Dunning, Joey Huchette, Miles Lubin, and contributors
```
Comment (mlubin) [inline src/constraints_future.jl]: Is this change related to the PR?

Comment (blegat) [reply src/constraints_future.jl]: Ups it shouldn't have been pushed
